### PR TITLE
Find all unique prime factor of any number n

### LINKED ISSUE #3005: All unique prime factors of n (in C++)
## 💥 Proposal

(A clear and concise description of what the proposal is.)

### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#reporting-new-issues)?

(Write your answer here.)
Yes , I have read the contributing guidelines.
Hey , I am Kirti as a GSSoc'21 participant,
In this algorithm  we can find all unique prime factors of any number n where n>1.For example if n is 24 then output will be (2,3), if  n is 30 then out put will be (2,3,5).
Please assign this task to me!

### PR DESCRIPTION
**Pull request by Kirti1807**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

### Description

(Write your answer here.)

### Checklist

- [ Yes ] I've read the contribution guidelines.
- [Yes ] I've checked the issue list before deciding what to submit.
- [Yes ] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(Write your answer here.)
In this I try  to find all unique prime factors of any number n.
FIXES #3005

### CODE REVIEW

**Comment by Kirti1807** [comment]
@ankitaggarwal23 
Sir please review it.

**Comment by Kirti1807** [comment]
Please review it ,I am not able to understand why this tests were not passed.

File: C-Plus-Plus/README.md [hunk 1]
```diff
@@ -59,6 +59,12 @@
 - [Ticket Auction](cp/tkt_auction.cpp)
 - [Replace Pi using recursion](cp/replace_pi.cpp)
 
+
+
+
+
+
```
Comment (HarshCasper) [inline C-Plus-Plus/README.md]: Remove all these blank lines

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+   Test case_1 Input: 111
+   Test case_1 Output: 3  37
+
+   Test case_2 Input:5446
+   Test case_2 Output: 2  7  389
+*/
+/* Time complaxity : O(n*n)
+   Space Complaxity : O(n*n)
+*/
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Put a new line at the EOF

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
+//This is a use of Sieve of Eratosthenes algorethm.
+//In this We can find all unique prime factor of any number n.
+
+#include<bits/stdc++.h>
+using namespace std;
+int main(){
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Shift all your code logic into another function and use main() for driver code.

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
+//This is a use of Sieve of Eratosthenes algorethm.
+//In this We can find all unique prime factor of any number n.
+
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Document it properly. And don't make silly typos (`algorethm`)

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  for(int i = 1 ; i <= num ; i++) arr[i]=1;
+  map<int ,vector<int>> m;
+  //now store all prime factor in vector of respective key in map
+  for(int i=2 ; i <= num ; i++) {
+    if(arr[i] == 0)continue;
+    for(int j=i ; j <= num ; j=j+i){
+      m[j].push_back(i);
+      arr[j]=0;
+    }
+  }
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Mess of a code. Make it less complicated and format/indent it

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by Kirti1807** [comment]
Sir, I try to make all required changes. Please review it again

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+   Test case_1 Input: 111
+   Test case_1 Output: 3  37
+
+   Test case_2 Input:5446
+   Test case_2 Output: 2  7  389
+*/
+/* Time complaxity : O(n*n)
+   Space Complaxity : O(n*n)
+*/
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Add a new line at the EOF

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
+//This is a use of Sieve of Eratosthenes algorithm.
+//In this We can find all unique prime factor of any number n.
+
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Put a proper description

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
+//This is a use of Sieve of Eratosthenes algorithm.
+//In this We can find all unique prime factor of any number n.For example if n is 12 then output will be 2,3
+//In this first we store all prime factor of numbers till n and then print prime factors of n.
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Put in a multi-line comment

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+#include<bits/stdc++.h>
+using namespace std;
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp:6]: Add a new line after using namespace std;

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      m[j].push_back(i);
+      arr[j]=0;
+    }
+  }
+  //print vector when key is n
+  for(int i=0 ; i < (m[num].size()) ; i++){
+    cout<<m[num][i]<<"  ";
+  }
+  return;
+}
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp:28]: Add a new line above the int main()

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+/* Sample Input: 24
+   Sample output: 2  3
+
+   Test case_1 Input: 111
+   Test case_1 Output: 3  37
+
+   Test case_2 Input:5446
+   Test case_2 Output: 2  7  389
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Remove the extra sample cases one is enough.
Capture

Kindly display a proper message using `cout` while reading and writing I/P and O/P

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  int n;
+  cout<<"Enter the number: ";
+  cin>>n;
+  unique_prime(n);
+  
+}
+
+/* Sample Input:5446
+   Sample output: 2  7  389
+*/
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Capture

Your code produces I/P & O/P like this kindly change it.

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Comment by siddharth25pandey** [comment]
@Kirti1807 Please store the Prime Factors in **set** and reduce the Time Complexity for finding the Unique Prime factor

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  map<int ,vector<int>> m;
+  //now store all prime factor in vector of respective key in map
+  for(int i=2 ; i <= num ; i++) {
+    if(arr[i] == 0) {
+      continue;
+    }
+    for(int j=i ; j <= num ; j=j+i) {
+      m[j].push_back(i);
+      arr[j]=0;
+    }
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: @Kirti1807 Please store the Prime Factors in set and reduce the Time Complexity for finding the Unique Prime factor

**Review by siddharth25pandey — CHANGES_REQUESTED** [review]
@Kirti1807 Do the suggested changes

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+In this first we store all prime factor of numbers till n and then print prime factors of n. */
+
+#include <bits/stdc++.h>
+using namespace std;
+
+void unique_prime(int n){
+  int arr[n+1];
+  for(int i = 2 ; i <= n ; i++ ) {
+    arr[i]=0;
+  }
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Remove the Redundant line, Directly use **memset** for initializing the array

**Review by siddharth25pandey — CHANGES_REQUESTED** [review]
@Kirti1807 Do the suggested changes or if possible
Try to create a **SET**  not **ARRAY** for storing prime numbers uniquely

File: C-Plus-Plus/cp/Unique_prime_factors.cpp [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        arr[j] = 1;
+      }
+    }
+  }
+  //now check which prime number are factor of n
+  cout<<"All unique prime factors of "<<n<<" are: ";
+  for(int i=2; i<=n; i++){  
+    if(arr[i] == 0 && n%i == 0 ){cout << i <<"  ";}
+  }
+}
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Unique_prime_factors.cpp]: Rewrite it to
```
if(arr[i] == 0 && n%i == 0 ){
        cout << i <<"  ";
     }
  }
}
```

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Review by siddharth25pandey — APPROVED** [review]
@LGTM 💯 Able to merge

**Comment by siddharth25pandey** [comment]
@all-contributors please add @Kirti1807 for her code

**Pull request merged**